src/Hpack/Syntax/Dependency.hs

```diff
@@ -34,33 +39,92 @@ instance FromValue Dependencies where
     Object _ -> Dependencies <$> fromValue v
     _ -> typeMismatch "Array, Object, or String" v
     where
-      fromDependency :: Dependency -> (String, DependencyVersion)
-      fromDependency (Dependency name version) = (name, version)
+      fromDependency :: Dependency -> (String, DependencyInfo)
+      fromDependency (Dependency name info) = (name, info)
 
       dependenciesFromList :: [Dependency] -> Dependencies
       dependenciesFromList = Dependencies . Map.fromList . map fromDependency
 
+data ModuleExposure
+  = ModuleDefaultExpose { _hiding :: [String] }
+  | ModuleDefaultHide { _exposing :: [String] }
+  deriving (Eq, Show)
+
+defaultModuleExposure :: ModuleExposure
+defaultModuleExposure = ModuleDefaultExpose []
+
+moduleExposureFromObject :: Object -> Parser ModuleExposure
+moduleExposureFromObject o = do
+  exposingMay <- o .:? "exposing"
+  hidingMay <- o .:? "hiding"
+  case (exposingMay, hidingMay) of
+    (Nothing, Nothing) ->
+      return defaultModuleExposure
+    (Just exposing, Nothing) ->
+      return $ ModuleDefaultHide exposing
+    (Nothing, Just hiding) ->
+      return $ ModuleDefaultExpose hiding
+    (Just _, Just _) ->
+      fail "Cannot combine 'exposing' and 'hiding' in a single dependency."
+
+data ModuleRenaming = ModuleRenaming {
+  moduleRenamingFrom :: String
+, _moduleRenamingTo :: String
+} deriving (Eq, Show)
+
+instance FromValue ModuleRenaming where
+  fromValue = withObject $ \ o ->
+    ModuleRenaming <$> (o .: "from") <*> (o .: "to")
+
+data DependencyInfo = DependencyInfo {
+  dependencyVersion :: DependencyVersion
+, dependencySourceDependency :: Maybe SourceDependency
+, dependencyExposure :: ModuleExposure
+, dependencyRenaming :: [ModuleRenaming]
+} deriving (Eq, Show)
+
+dependencyInfoFromObject :: Object -> Parser DependencyInfo
+dependencyInfoFromObject o = do
+  version <- o .: "version" <|> return AnyVersion
+  sourceDepMay <-
+       (Just <$> sourceDependencyFromObject o)
+    <|> return Nothing
+  exposure <- moduleExposureFromObject o
+  renaming <- o .: "renaming" <|> return []
+  return $ DependencyInfo version sourceDepMay exposure renaming
+
+instance FromValue DependencyInfo where
+  fromValue v = case v of
+    Null -> return $ DependencyInfo AnyVersion Nothing defaultModuleExposure []
+    Number n -> do
+      let version = scientificToDependencyVersion n
+      return $ DependencyInfo version Nothing defaultModuleExposure []
+    String s -> do
+      let input = T.unpack s
+      version <- parseVersionRange ("== " ++ input) <|> parseVersionRange input
+      return $ DependencyInfo version Nothing defaultModuleExposure []
+    Object o -> dependencyInfoFromObject o
+    _ -> typeMismatch "Null, Object, Number, or String" v
```

Comment:
Ideally I'd like to share more code with the `FromValue DependencyVersion` instance, but I can't quite get it to work without regressing the error message here.